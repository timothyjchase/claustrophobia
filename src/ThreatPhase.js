import React, { Component } from 'react'
import { observer } from 'mobx-react'
import { Button, Checkbox, Icon, Input, Message } from 'semantic-ui-react'
import EventMessage from './EventMessage'

const ThreatEventStep = ({ game }) => (
  <div>
    <EventMessage game={game} />
    <Button.Group vertical fluid>
      <Button primary onClick={() => game.completeThreatEventStep()}>
        <Icon name="play" /> Next
      </Button>
    </Button.Group>
  </div>
)

const CheckLegalPlacementStep = ({ game }) => (
  <div>
    <Checkbox
      label={`At least 1 legal placement for Demon warriors within 4 tiles of
        a Human warrior?`}
      onChange={() => (game.legalPlacement = !game.legalPlacement)}
      checked={game.legalPlacement}
    />
    <br />
    <br />
    <Button.Group vertical fluid>
      <Button primary onClick={() => game.completeThreatLegalPlacementStep()}>
        <Icon name="play" /> Next
      </Button>
    </Button.Group>
  </div>
)

const SpawnDemonStep = ({ game }) => (
  <div>
    {!game.legalPlacement && (
      <Message negative>
        <Message.Header>A Taste for Blood</Message.Header>
        <Message.Content>
          Human warriors cannot prevent the appearance of Demons.
        </Message.Content>
      </Message>
    )}
    {game.scenario !== 'THE_SURVIVORS' &&
      game.scenario !== 'THE_RITUAL' && (
        <p>
          Add the <strong>{game.getDemon().name}</strong> to the unexplored
          opening closest to the tile with the largest group of humans.
        </p>
      )}
    {game.scenario === 'THE_SURVIVORS' && (
      <p>
        Add the <strong>{game.getDemon().name}</strong> to the unexplored
        opening closest to the tile with Fresh Air.
      </p>
    )}
    {game.scenario === 'THE_RITUAL' && (
      <p>
        Add the <strong>{game.getDemon().name}</strong> to the unexplored
        opening farthest from a Human warrior.
      </p>
    )}
    {game.legalPlacement && (
      <p>
        <i>Reminder: Human warriors block tiles</i>
      </p>
    )}
    <Button.Group vertical fluid>
      <Button primary onClick={() => game.completeThreatSpawnDemonStep()}>
        <Icon name="play" /> Next Phase
      </Button>
    </Button.Group>
  </div>
)

class SpawnTrogsStep extends Component {
  constructor(props) {
    super(props)
    const maxTrogs = props.game.legalPlacement
      ? props.game.threatDice
      : Math.min(props.game.threatDice, 3)
    this.state = { trogsAdded: maxTrogs, maxTrogs }
  }

  render = () => {
    const { game } = this.props
    const { trogsAdded, maxTrogs } = this.state
    return (
      <div>
        {!game.legalPlacement && (
          <Message negative>
            <Message.Header>A Taste for Blood</Message.Header>
            <Message.Content>
              Human warriors cannot prevent the appearance of Troglodytes.
            </Message.Content>
          </Message>
        )}
        {game.scenario !== 'THE_SURVIVORS' &&
          game.scenario !== 'THE_RITUAL' && (
            <p>
              Add up to <strong>{maxTrogs}</strong> Troglogytes to unexplored
              openings closest to the tile with the largest group of humans.
            </p>
          )}
        {game.scenario === 'THE_SURVIVORS' && (
          <p>
            Add up to <strong>{maxTrogs}</strong> Troglogytes to unexplored
            openings closest to the tile with Fresh Air.
          </p>
        )}
        {game.scenario === 'THE_RITUAL' && (
          <p>
            Add up to <strong>{maxTrogs}</strong> Troglogytes to unexplored
            openings with the following priority:
          </p>
        )}
        {game.scenario === 'THE_RITUAL' && (
          <ul>
            <li>Seal of Protection on its path to the Pentacle Room</li>
            <li>Fewest Human warriors on its path to the Pentacle Room </li>
          </ul>
        )}
        {game.legalPlacement && (
          <p>
            <i>Reminder: Human warriors block tiles</i>
          </p>
        )}
        <Input
          label="Troglogytes added"
          placeholder="0"
          type="number"
          min={0}
          max={6}
          value={trogsAdded}
          onChange={(event, { value }) => this.setState({ trogsAdded: value })}
          fluid
        />
        <br />
        <Button.Group vertical fluid>
          <Button
            primary
            disabled={!trogsAdded}
            onClick={() =>
              game.completeThreatSpawnTrogsStep(parseInt(trogsAdded))
            }
          >
            <Icon name="play" /> Next
          </Button>
        </Button.Group>
      </div>
    )
  }
}

const CheckTrogsDistanceStep = ({ game }) => (
  <div>
    <Checkbox
      label="Any Troglogytes 2 or more tiles away from the closest human?"
      onChange={() => (game.trogsFar = !game.trogsFar)}
      checked={game.trogsFar}
    />
    <br />
    <br />
    <Checkbox
      label="Any Troglogytes adjacent to or on the same tile as a human?"
      onChange={() => (game.trogsClose = !game.trogsClose)}
      checked={game.trogsClose}
    />
    <br />
    <br />
    <Button.Group vertical fluid>
      <Button primary onClick={() => game.completeThreatPhase()}>
        <Icon name="play" /> Next Phase
      </Button>
    </Button.Group>
  </div>
)

const STEPS_CONFIG = {
  THREAT_EVENT: { component: observer(ThreatEventStep) },
  CHECK_LEGAL_PLACEMENT: { component: observer(CheckLegalPlacementStep) },
  SPAWN_DEMON: { component: observer(SpawnDemonStep) },
  SPAWN_TROGS: { component: observer(SpawnTrogsStep) },
  CHECK_TROGS_DISTANCE: { component: observer(CheckTrogsDistanceStep) },
}

const ThreatPhase = ({ game }) => {
  const StepComponent = STEPS_CONFIG[game.threatStep].component
  return (
    <div>
      <StepComponent game={game} />
    </div>
  )
}

export default observer(ThreatPhase)
