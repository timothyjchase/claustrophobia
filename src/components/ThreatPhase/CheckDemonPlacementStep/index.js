import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Button, Checkbox } from 'semantic-ui-react'

class CheckDemonPlacementStep extends Component {
  constructor(props) {
    super(props)
    this.state = { legalPlacement: false }
  }

  render() {
    const { completeThreatDemonPlacementStep } = this.props
    const { legalPlacement } = this.state

    return (
      <div>
        <Checkbox
          label="At least 1 legal placement for a Demon within 4 tiles of a Human warrior?"
          onChange={() => this.setState({ legalPlacement: !legalPlacement })}
          checked={legalPlacement}
        />
        <br />
        <br />
        <Button.Group vertical fluid>
          <Button
            primary
            icon="play"
            content="Next"
            onClick={() => completeThreatDemonPlacementStep(legalPlacement)}
          />
        </Button.Group>
      </div>
    )
  }
}

CheckDemonPlacementStep.propTypes = {
  completeThreatDemonPlacementStep: PropTypes.func.isRequired,
}

export default CheckDemonPlacementStep
