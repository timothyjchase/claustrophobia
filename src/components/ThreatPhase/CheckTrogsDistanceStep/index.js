import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Button, Checkbox } from 'semantic-ui-react'

class CheckTrogsDistanceStep extends Component {
  constructor(props) {
    super(props)
    this.state = { trogsFar: false, trogsClose: false }
  }

  render() {
    const { completeThreatTrogsDistanceStep } = this.props
    const { trogsFar, trogsClose } = this.state

    return (
      <div>
        <Checkbox
          id="trogsFar"
          label="Any Troglogytes 2 or more tiles away from the closest human?"
          onChange={() => this.setState({ trogsFar: !trogsFar })}
          checked={trogsFar}
        />
        <br />
        <br />
        <Checkbox
          id="trogsClose"
          label="Any Troglogytes adjacent to or on the same tile as a human?"
          onChange={() => this.setState({ trogsClose: !trogsClose })}
          checked={trogsClose}
        />
        <br />
        <br />
        <Button.Group vertical fluid>
          <Button
            primary
            icon="play"
            content="Next Phase"
            onClick={() =>
              completeThreatTrogsDistanceStep(trogsFar, trogsClose)
            }
          />
        </Button.Group>
      </div>
    )
  }
}

CheckTrogsDistanceStep.propTypes = {
  completeThreatTrogsDistanceStep: PropTypes.func.isRequired,
}

export default CheckTrogsDistanceStep
