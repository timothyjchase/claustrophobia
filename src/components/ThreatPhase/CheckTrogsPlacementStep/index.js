import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Button, Checkbox, Icon } from 'semantic-ui-react'

class CheckTrogsPlacementStep extends Component {
  constructor(props) {
    super(props)
    this.state = { legalPlacement: false }
  }

  render() {
    const { completeThreatTrogsPlacementStep } = this.props
    const { legalPlacement } = this.state

    return (
      <div>
        <Checkbox
          label={`At least 1 legal placement for Troglodytes within 4 tiles of
        a Human warrior?`}
          onChange={() => this.setState({ legalPlacement: !legalPlacement })}
          checked={legalPlacement}
        />
        <br />
        <br />
        <Button.Group vertical fluid>
          <Button
            primary
            onClick={() => completeThreatTrogsPlacementStep(legalPlacement)}
          >
            <Icon name="play" /> Next
          </Button>
        </Button.Group>
      </div>
    )
  }
}

CheckTrogsPlacementStep.propTypes = {
  completeThreatTrogsPlacementStep: PropTypes.func.isRequired,
}

export default CheckTrogsPlacementStep
