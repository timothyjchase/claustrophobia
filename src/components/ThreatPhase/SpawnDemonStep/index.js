import React from 'react'
import PropTypes from 'prop-types'
import renderHTML from 'react-render-html'
import { Button, Message } from 'semantic-ui-react'

const SpawnDemonStep = ({
  scenario,
  demonName,
  legalPlacement,
  completeThreatSpawnDemonStep,
}) => {
  const spawnLocation =
    scenario.demonSpawnLocation ||
    '<li>the <strong>largest group</strong> of Human warriors</li>'
  return (
    <div>
      {!legalPlacement && (
        <Message negative>
          <Message.Header>A Taste for Blood</Message.Header>
          <Message.Content>
            Human warriors cannot prevent the appearance of Demons.
          </Message.Content>
        </Message>
      )}
      <p>
        Add the
        <strong>{` ${demonName} `}</strong>
        to the tile closest to:
      </p>
      <ul>{renderHTML(spawnLocation)}</ul>
      {legalPlacement && (
        <p>
          <i>Reminder: Human warriors block tiles with unexplored openings</i>
        </p>
      )}
      <Button.Group vertical fluid>
        <Button
          primary
          icon="play"
          content="Next Phase"
          onClick={completeThreatSpawnDemonStep}
        />
      </Button.Group>
    </div>
  )
}

SpawnDemonStep.defaultProps = {
  demonName: 'Demon',
}

SpawnDemonStep.propTypes = {
  scenario: PropTypes.object.isRequired,
  demonName: PropTypes.string,
  legalPlacement: PropTypes.bool.isRequired,
  completeThreatSpawnDemonStep: PropTypes.func.isRequired,
}

export default SpawnDemonStep
