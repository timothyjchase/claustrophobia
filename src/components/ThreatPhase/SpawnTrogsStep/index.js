import React, { Component } from 'react'
import PropTypes from 'prop-types'
import renderHTML from 'react-render-html'
import { Button, Input, Message } from 'semantic-ui-react'

class SpawnTrogsStep extends Component {
  constructor(props) {
    super(props)
    const { legalPlacement, threatDice } = props
    const maxTrogs = legalPlacement ? threatDice : Math.min(threatDice, 3)
    this.state = { trogsAdded: maxTrogs, maxTrogs }
  }

  render = () => {
    const {
      scenario,
      legalPlacement,
      completeThreatSpawnTrogsStep,
    } = this.props
    const { trogsAdded, maxTrogs } = this.state
    const spawnLocation =
      scenario.trogsSpawnLocation ||
      '<li>the <strong>largest group</strong> of Human warriors</li>'

    return (
      <div>
        {!legalPlacement && (
          <Message negative>
            <Message.Header>A Taste for Blood</Message.Header>
            <Message.Content>
              Human warriors cannot prevent the appearance of Troglodytes.
            </Message.Content>
          </Message>
        )}
        <p>
          Add up to
          <strong>{` ${maxTrogs} `}</strong>
          Troglogytes to tiles closest to:
        </p>
        <ul>{renderHTML(spawnLocation)}</ul>
        {legalPlacement && (
          <p>
            <i>Reminder: Human warriors block tiles with unexplored openings</i>
          </p>
        )}
        <Input
          label="Troglogytes added"
          placeholder="0"
          type="number"
          min={0}
          max={6}
          value={trogsAdded}
          onChange={(event, { value }) => this.setState({ trogsAdded: value })}
          fluid
        />
        <br />
        <Button.Group vertical fluid>
          <Button
            primary
            disabled={!trogsAdded}
            icon="play"
            content="Next"
            onClick={() =>
              completeThreatSpawnTrogsStep(parseInt(trogsAdded, 10))
            }
          />
        </Button.Group>
      </div>
    )
  }
}

SpawnTrogsStep.propTypes = {
  scenario: PropTypes.object.isRequired,
  legalPlacement: PropTypes.bool.isRequired,
  threatDice: PropTypes.number.isRequired,
  completeThreatSpawnTrogsStep: PropTypes.func.isRequired,
}

export default SpawnTrogsStep
