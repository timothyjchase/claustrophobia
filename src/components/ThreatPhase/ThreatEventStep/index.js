import React from 'react'
import PropTypes from 'prop-types'
import { Button, Icon } from 'semantic-ui-react'

const ThreatEventStep = ({ completeThreatEventStep }) => (
  <div>
    <Button.Group vertical fluid>
      <Button primary onClick={completeThreatEventStep}>
        <Icon name="play" /> Next
      </Button>
    </Button.Group>
  </div>
)

ThreatEventStep.propTypes = {
  completeThreatEventStep: PropTypes.func.isRequired,
}

export default ThreatEventStep
