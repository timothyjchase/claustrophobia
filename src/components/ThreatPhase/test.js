import React from 'react'
import ThreatPhase from './index'

describe('<ThreatPhase />', () => {
  const commonProps = {
    scenario: {},
    threatDice: 6,
    legalPlacement: true,
    completeThreatDemonPlacementStep: jest.fn(),
    completeThreatTrogsDistanceStep: jest.fn(),
    completeThreatTrogsPlacementStep: jest.fn(),
    completeThreatSpawnDemonStep: jest.fn(),
    completeThreatSpawnTrogsStep: jest.fn(),
    completeThreatEventStep: jest.fn(),
  }
  describe('with step = CHECK_DEMON_PLACEMENT', () => {
    const component = shallow(
      <ThreatPhase threatStep="CHECK_DEMON_PLACEMENT" {...commonProps} />,
    )
    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
  describe('with step = CHECK_TROGS_DISTANCE', () => {
    const component = shallow(
      <ThreatPhase threatStep="CHECK_TROGS_DISTANCE" {...commonProps} />,
    )
    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
  describe('with step = CHECK_TROGS_PLACEMENT', () => {
    const component = shallow(
      <ThreatPhase threatStep="CHECK_TROGS_PLACEMENT" {...commonProps} />,
    )
    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
  describe('with step = SPAWN_DEMON', () => {
    const component = shallow(
      <ThreatPhase threatStep="SPAWN_DEMON" {...commonProps} />,
    )
    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
  describe('with step = SPAWN_TROGS', () => {
    const component = shallow(
      <ThreatPhase threatStep="SPAWN_TROGS" {...commonProps} />,
    )
    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
  describe('with step = THREAT_EVENT', () => {
    const component = shallow(
      <ThreatPhase threatStep="THREAT_EVENT" {...commonProps} />,
    )
    it('should match snapshot', () => {
      expect(component).toMatchSnapshot()
    })
  })
})
