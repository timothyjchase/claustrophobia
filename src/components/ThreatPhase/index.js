import React from 'react'
import PropTypes from 'prop-types'
import ThreatEventStep from './ThreatEventStep'
import CheckDemonPlacementStep from './CheckDemonPlacementStep'
import CheckTrogsPlacementStep from './CheckTrogsPlacementStep'
import SpawnDemonStep from './SpawnDemonStep'
import SpawnTrogsStep from './SpawnTrogsStep'
import CheckTrogsDistanceStep from './CheckTrogsDistanceStep'

const STEPS_CONFIG = {
  CHECK_DEMON_PLACEMENT: CheckDemonPlacementStep,
  CHECK_TROGS_DISTANCE: CheckTrogsDistanceStep,
  CHECK_TROGS_PLACEMENT: CheckTrogsPlacementStep,
  SPAWN_DEMON: SpawnDemonStep,
  SPAWN_TROGS: SpawnTrogsStep,
  THREAT_EVENT: ThreatEventStep,
}

const ThreatPhase = ({
  scenario,
  threatStep,
  demonName,
  threatDice,
  legalPlacement,
  completeThreatDemonPlacementStep,
  completeThreatTrogsDistanceStep,
  completeThreatTrogsPlacementStep,
  completeThreatSpawnDemonStep,
  completeThreatSpawnTrogsStep,
  completeThreatEventStep,
}) => {
  const StepCompoonent = STEPS_CONFIG[threatStep]
  let stepProps = {}
  if (threatStep === 'CHECK_DEMON_PLACEMENT') {
    stepProps = { completeThreatDemonPlacementStep }
  } else if (threatStep === 'CHECK_TROGS_DISTANCE') {
    stepProps = { completeThreatTrogsDistanceStep }
  } else if (threatStep === 'CHECK_TROGS_PLACEMENT') {
    stepProps = { completeThreatTrogsPlacementStep }
  } else if (threatStep === 'SPAWN_DEMON') {
    stepProps = {
      scenario,
      demonName,
      legalPlacement,
      completeThreatSpawnDemonStep,
    }
  } else if (threatStep === 'SPAWN_TROGS') {
    stepProps = {
      scenario,
      threatDice,
      legalPlacement,
      completeThreatSpawnTrogsStep,
    }
  } else if (threatStep === 'THREAT_EVENT') {
    stepProps = { completeThreatEventStep }
  }
  return <StepCompoonent {...stepProps} />
}

ThreatPhase.defaultProps = {
  demonName: 'Demon',
}

ThreatPhase.propTypes = {
  scenario: PropTypes.object.isRequired,
  threatStep: PropTypes.string.isRequired,
  demonName: PropTypes.string,
  threatDice: PropTypes.number.isRequired,
  legalPlacement: PropTypes.bool.isRequired,
  completeThreatDemonPlacementStep: PropTypes.func.isRequired,
  completeThreatTrogsDistanceStep: PropTypes.func.isRequired,
  completeThreatTrogsPlacementStep: PropTypes.func.isRequired,
  completeThreatSpawnDemonStep: PropTypes.func.isRequired,
  completeThreatSpawnTrogsStep: PropTypes.func.isRequired,
  completeThreatEventStep: PropTypes.func.isRequired,
}

export default ThreatPhase
