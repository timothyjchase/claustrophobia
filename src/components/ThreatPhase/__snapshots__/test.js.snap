// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThreatPhase /> with step = CHECK_DEMON_PLACEMENT should match snapshot 1`] = `
<CheckDemonPlacementStep
  completeThreatDemonPlacementStep={[MockFunction]}
/>
`;

exports[`<ThreatPhase /> with step = CHECK_TROGS_DISTANCE should match snapshot 1`] = `
<CheckTrogsDistanceStep
  completeThreatTrogsDistanceStep={[MockFunction]}
/>
`;

exports[`<ThreatPhase /> with step = CHECK_TROGS_PLACEMENT should match snapshot 1`] = `
<CheckTrogsPlacementStep
  completeThreatTrogsPlacementStep={[MockFunction]}
/>
`;

exports[`<ThreatPhase /> with step = SPAWN_DEMON should match snapshot 1`] = `
<SpawnDemonStep
  completeThreatSpawnDemonStep={[MockFunction]}
  demonName="Demon"
  legalPlacement={true}
  scenario={Object {}}
/>
`;

exports[`<ThreatPhase /> with step = SPAWN_TROGS should match snapshot 1`] = `
<SpawnTrogsStep
  completeThreatSpawnTrogsStep={[MockFunction]}
  legalPlacement={true}
  scenario={Object {}}
  threatDice={6}
/>
`;

exports[`<ThreatPhase /> with step = THREAT_EVENT should match snapshot 1`] = `
<ThreatEventStep
  completeThreatEventStep={[MockFunction]}
/>
`;
